torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -264,6 +280,53 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
   replica.contents = grad;
 }
 
+// The function `autograd_hook` is called after the gradient for a
+// model parameter has been accumulated into its gradient tensor.
+// This function is only to be called from the autograd thread.
+// Comared to the vinilla autograd_hook() function, this delayed hook installs
+// a final callback to the autograd engine, and launched all communication in
+// the final callback.
+void Reducer::delayed_autograd_hook(VariableIndex index) {
+  std::lock_guard<std::mutex> lock(this->mutex_);
+
+  // Since it gets here, this param has been used for this iteration. We want
+  // to mark it in local_used_maps_. During no_sync session, the same var can
+  // be set multiple times, which is OK as does not affect correctness. As long
+  // as it is used once during no_sync session, it is marked as used.
+  local_used_maps_[index.replica_index][index.variable_index] = 1;
+
+  // Ignore if we don't expect to be called.
+  // This may be the case if the user wants to accumulate gradients
+  // for number of iterations before reducing them.
+  if (!expect_autograd_hooks_) {
+    return;
+  }
+
+  if (require_final_hook_) {
+    torch::autograd::Engine::get_default_engine().queue_callback([=] {
+      std::lock_guard<std::mutex> lock(this->mutex_);
+
+      const auto replica_count = replicas_.size();
+      grad_accumulators_.resize(replica_count);
+      for (size_t replica_index = 0; replica_index < replica_count;
+           replica_index++) {
+        const auto variable_count = replicas_[replica_index].size();
+        for (size_t variable_index = 0; variable_index < variable_count;
+             variable_index++) {
+          const auto index = VariableIndex{
+              .replica_index = replica_index,
+              .variable_index = variable_index,
+          };
+
+          mark_variable_ready(index);
```

Comment:
I'm wondering what is the performance difference comparing to just directly call all reduce to all gradients here? The main benefit to do bucket all reduce here probably is that it can work with round robin process groups.